Mechanism for Mono to Clone Swift PM Repos

tjprescott opened the pull request:
The goal here is for us to have fine-grained, independent version control for SwiftPM modules the way we do in other SDKs and with CocoaPods while maintaining the benefits of a mono-repo.

To facilitate this, we will have a per-module Package.swift and CHANGELOG.md. Each module will be represented in `manifest.yml` with its path in the mono-repo and its mirror repo name.

Whenever the mono-repo is updated, we can update the mirror repo with:

`python3 eng/scripts/sync_repo.py <MODULE>`

This will copy that branch to the root of the mirror, which can then be target by SwiftPM. In the current model, users target the "AzureSDK" package, and then specify individual modules, whereas with this change, the target will be the module.

The current example of this is AzureCore, which will mirror to `azure-sdk-for-ios-core`.

weshaggard commented on Package.swift:
```diff
@@ -35,7 +35,6 @@ let package = Package(
         .macOS(.v10_15), .iOS(.v12), .tvOS(.v12)
     ],
     products: [
```
Don't we want to remove this root swift file completely?

tjprescott replied:
Yes, we will. At that point, no one would be able to target this repo with SPM. For this PR, I'm only separating out AzureCore to see what that would look like.

weshaggard replied:
OK once we do move over I expect folks can still use an older release tag for the preview packages we did ship from the repo still.

tjprescott replied:
Yes, the old release tags will work

tjprescott replied:
Removed the root Package.swift.

weshaggard commented on manifest.yaml:
```diff
@@ -0,0 +1,29 @@
+AzureCore:
```
What is this file used for? Ideally we should try to avoid yet another list of projects and versions to maintain if we can.

tjprescott replied:
Right now the script only uses the source and mirror entries. Unless we have a deterministic way of mapping SDKs to repos, I don't see how we will avoid this. `Package.swift` files don't actually specify versions, so the idea behind version would be a simple place to acquire the correct version for tagging. We could potentially pull that from the podspec though.

weshaggard replied:
Ideally we would only have one place that defines the version. If swift doesn't need it perhaps we could use the pod. I know some languages also embed the version in source code so it can also be used by code if needed.

weshaggard replied:
As for the other metadata here could we gather it from the swift files instead of explicitly listing it? For most languages we essentially traverse the sdk looking for some sort of project file and then read the metadata from there.

mitchdenny commented:
This PR seems to have two strategies in play. One strategy that has been discussed is having a copy of all the modules in a separate repository (synchronized from the mono-repo). But there also appears to be a reference to an ```azure-sdk-for-ios-core``` repository as well (which makes me think that this is just breaking out core separately, and will be managed as a separate repo (incl the engineering system etc).

tjprescott commented:
> This PR seems to have two strategies in play. One strategy that has been discussed is having a copy of all the modules in a separate repository (synchronized from the mono-repo). But there also appears to be a reference to an azure-sdk-for-ios-core repository as well (which makes me think that this is just breaking out core separately, and will be managed as a separate repo (incl the engineering system etc).

@mitchdenny thanks for the catch. I updated the reference to `SwiftPM-AzureCore`. Also, @weshaggard I updated the `sync_repo.py` script so it no longer requires any metadata file.

tjprescott commented:
@weshaggard @mitchdenny this PR should be ready for review.

mitchdenny reviewed (commented):
One comment around the copy logic. It would really need to delete-then-copy the contents of the directory otherwise you won't be catching renames etc.

Another thing that needs to be worked out is the tagging strategy when a release goes out. Is the plan just to replace the contents of the repo and then commit and tag? So we could end up with something like this:

commit1: 1.0.0-beta.11
commit2: 1.0.0
commit3: 1.1.0-beta.1
commit4: 1.0.1

... just based on blindly doing a copy/tag. It probably doesn't make much difference since folks should always be going from a tag in the swiftpm-* repos, but if anyone looked at the history of the repo it would be totally confusing. To try and mitigate that I think it might make sense to pre-pend something in the README.md file which provides a back-pointer to the actual commit in the mono-repo so folks know to navigate back there to find the actual state of the universe.

I think there will also need to be some content in the README that explains how/why these repos are done this way. Imagine a scenario where someone inherits an iOS code-base and they go and look at their dependencies and they see one of our swiftpm-* links. They may not have previously worked with the Azure SDK before so wouldn't know why we have these satellite repos.

tjprescott commented on eng/scripts/sync_repo.py:
```diff
@@ -0,0 +1,75 @@
+        _log_error_and_quit(f'Destination path does not exist: {dest_path}')
+
+    # copy the subfolder to the root of the SwiftPM repo
+    _copy(source_path, dest_path)
```
Good catch. I have updated the script so it will clear the folder before copying.

tjprescott commented:
> Another thing that needs to be worked out is the tagging strategy when a release goes out. Is the plan just to replace the contents of the repo and then commit and tag? So we could end up with something like this:
> 
> commit1: 1.0.0-beta.11
> commit2: 1.0.0
> commit3: 1.1.0-beta.1
> commit4: 1.0.1

There would be corresponding tags on the monorepo like:

commit1: AzureCore_1.0.0-beta.11
commit2: AzureCore_1.0.0 (albeit not anytime soon)
commit3: AzureCore_1.1.0-beta.1
commit4: AzureCore_1.0.1

However, if that were the sequence in both repos, I don't see how either would be any less confusing. \

> I think there will also need to be some content in the README that explains how/why these repos are done this way. Imagine a scenario where someone inherits an iOS code-base and they go and look at their dependencies and they see one of our swiftpm-* links. They may not have previously worked with the Azure SDK before so wouldn't know why we have these satellite repos.

I think, most likely, they wouldn't think twice about the SwiftPM repo but would be surprised to learn of the existence of the mono-repo. There would be no reference to `azure-sdk-for-ios` in their files, but if they wanted to file a bug or contribute, the README would direct them to the mono-repo. But you are right, we should call this out in some fashion somewhere.

weshaggard commented on sdk/communication/AzureCommunication/Package.swift:
```diff
@@ -0,0 +1,69 @@
+        // Build targets
+        .target(
+            name: "AzureCommunication",
+            dependencies: ["AzureCore"],
```
I don't understand swift packages that much but don't you need to list a version and/or repository? Does this essentially take a source level dependency the way it is currently configured?

mitchdenny replied:
No the git reference is the version. Swift is 100% tied to Git as a version control mechanism.

tjprescott replied:
The target dependencies reference the package dependencies by name. The package dependencies will reference the repos (in this case, the SwiftPM-* repos) and either a version, commit, or branch.

weshaggard replied:
Where does that repo name and tag info get stored for these dependencies?

weshaggard replied:
I see that information listed under the dependencies section above now.

tjprescott replied:
Here's the example from Chat

let package = Package(
    name: "AzureCommunicationChat",
    platforms: [
        .iOS(.v12)
    ],
    products: [
        .library(name: "AzureCommunicationChat", targets: ["AzureCommunicationChat"])
    ],
    dependencies: [
        **.package(name: "AzureCore", url: "https://github.com/Azure/SwiftPM-AzureCore.git", .branch("master")),**
    ],
    targets: [
        // Build targets
        .target(
            name: "AzureCommunicationChat",
            **dependencies: ["AzureCore"],**
            path: "Source",
            exclude: [
                "README.md",
                "Tests",
                "Source/Supporting Files",
                "LICENSE"
            ]
        ),
...

weshaggard commented on sdk/communication/AzureCommunication/Package.swift:
```diff
@@ -0,0 +1,69 @@
+import PackageDescription
+
+let package = Package(
+    name: "AzureCommunication",
```
Do the project files not contain any sort of version?

tjprescott replied:
No they do not.

weshaggard commented on eng/scripts/sync_repo.py:
```diff
@@ -0,0 +1,86 @@
+    _copy(source_path, dest_path)
+
+    # copy the LICENSE, CCONTRIBUTING.md, and .gitignore files
+    for fname in ["LICENSE", "CONTRIBUTING.md", ".gitignore"]:
```
On top of these did we plan to also mutate the readme in the root to point back to the mono repo? Also we should probably update the repo description with a pointer as well.

tjprescott replied:
So, I updated the READMEs so that they target the appropriate repos (either the SwiftPM-* ones or the mono repo), so I'm not sure we need to do any mutation. If we get a lot of people expressing confusion, I'll address the issue then.

weshaggard commented on Package.swift:
```diff
@@ -45,17 +44,6 @@ let package = Package(
-            ],
-            sources: ["Source"]
-        ),
         .target(
```
Shouldn't these also be removed now?

tjprescott replied:
Removed.

weshaggard commented on examples/AzureSDKSmokeTestSwiftPM/AzureSDKSmokeTestSwiftPM.xcodeproj/project.pbxproj:
```diff
@@ -0,0 +1,360 @@
+// !$*UTF8*$!
```
Where does this get build and ran? Also how does it actually pull in the swift package references? Perhaps I'm missing something but I would expect to pull them in somewhere.

tjprescott replied:
I think I had that in there for testing. For now I will remove it and re-add it in a later PR, as it should be part of CI the way the CocoaPods smoke test is.

weshaggard commented on sdk/communication/AzureCommunicationChat/Package.swift:
```diff
@@ -30,84 +30,49 @@
+        .package(
+            name: "AzureCommunication",
+            url: "https://github.com/Azure/SwiftPM-AzureCommunication.git",
+            .branch("master")
```
I assume at some point these will turn into a release tags over a branch reference to master.

weshaggard reviewed (approved):
Overall this looks like a good starting point. Once you finish your testing lets get it in and start iterating.
